trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js

```diff
@@ -0,0 +1,41 @@
+const { dealHasAllValidUkefIds } = require('./dealHasAllUkefIds');
+const { submitACBSIfAllPartiesHaveUrn } = require('../controllers/deal.controller');
+const CONSTANTS = require('../../constants');
+
+/**
+ * Submit's deal to ACBS through the means of verification.
+ * This function performs following checks:
+ * 1. Valid submission type either AIN or MIN
+ * 2. Valid first submission check, ensuring `tfm.acbs` property does not exist
+ * 3. Valid deal and facility IDs
+ * @param {Object} deal - TFM deal object
+ * @param {boolean} [firstSubmissionCheck=true] - Flag indicating whether to perform the first submission check, defaults to `true`
+ * @param {boolean} [validateOnly=false] - Flag indicating whether to just to perform validation, defaults to `false`
+ * @returns {boolean} - True if the deal can be submitted to ACBS, otherwise false.
+ */
+const submitDealToACBS = async (deal, firstSubmissionCheck = true, validateOnly = false) => {
```

Comment:
I really like the validation on this function.

Would it be best to split this into two different functions? Someone in a few years time might find it surprising to see `submitDealToAcbs(bla,true, true)` would not actually submit a deal to acbs I think?

What do you think about something like the below

```
const getIsValidDealToSubmit = async (deal, isFirstSubmission) => {
  console.info('⚡ Validating ACBS checks for deal %s', _id);

  const { _id, dealSnapshot, tfm } = deal;

  const acceptableSubmissionTypes = [CONSTANTS.DEALS.SUBMISSION_TYPE.AIN, CONSTANTS.DEALS.SUBMISSION_TYPE.MIN];
  const isValidSubmissionType = acceptableSubmissionTypes.includes(dealSnapshot?.submissionType);
  const isValidFirstSubmission = isFirstSubmission ? !tfm?.acbs : true;
  const isValidIds = await dealHasAllValidUkefIds(_id);
  const isValidDealToSubmit = isValidSubmissionType && isValidFirstSubmission && isValidIds;

  return isValidDealToSubmit;
};

const submitDealToACBS = async ({ deal, isFirstSubmission = true }) => {
  try {
    const { _id } = deal;
    const isValidDealToSubmit = getIsValidDealToSubmit(deal, isFirstSubmission);
    if (!isValidDealToSubmit) {
      throw new InvalidDealError({ dealId: _id }); // these could just be standard Errors
    }
    await submitACBSIfAllPartiesHaveUrn(_id);
  } catch (error) {
    console.error('Unable to submit deal to ACBS %o', error); // We shouldn't really need this -- it should log fine by just throwing the errors
    throw new UnableToSubmitDealToAcbsError({ dealId: _id, cause: error }); // these could just be standard Errors
  }
};
```

Reply:
Just a side note -- if we are defaulting `isFirstSubmission` to true, could we remove it entirely and add in a check where we call it when needed ie:

```
if(isFirstSubmission){
submitDealToACBS(deal)
}
```

Im honestly not very bothered either way, just thinking it might be nicer to do this check outside of the function unless we're doing it in multiple places

Reply:
There were two separate functions before, however I have added these checks on the submission to avoid a bug or an eventuality where submission to ACBS can be invoked without any pre-requesite checks.

Reply:
I think we can still have the verification called inside the submit deal function, but keeping it seperate means we can call it seperately too unless I'm misunderstanding something?

ie:

```
const validateDealCanBeSubmittedToAcbs = async (deal, isFirstSubmission) => {
  // Validation stuff
    if (!isValidDealToSubmit) {
      throw new InvalidDealError({ dealId: _id });
    }
};

const submitDealToACBS = async ({ deal, isFirstSubmission = true }) => {
  try {
    validateDealCanBeSubmittedToAcbs(deal, isFirstSubmission); // we call the validation check here
    // Other logic
  } catch (error) {
    throw new UnableToSubmitDealToAcbsError({ dealId: _id, cause: error });
  }
};
```



------------------------------------



One of the reasons I'm slightly worried is I find this in `deal.submit.controller.js` line 184 very hard to understand without understanding that `submitDealToACBS` does not actually `submitDealToAcbs` the second time it's called, which is surprising. The function is both being used for `submitDealToACBS` and `validateDealCanBeSubmittedToAcbs` which I feel are two distinct functions (`submitDealToAcbs` can call `VAlidateDealCanBeSubmittedToAcbs`, as shown above)...

`deal.submit.controller.js` line 184:
```
    await submitDealToACBS(tfmDeal);

    if (await submitDealToACBS(tfmDeal, false, true)) {
      await issueAcbsFacilities(dealUpdate);
    }

    return tfmDeal;
```

Instead we could have:

```
try {
    await submitDealToAcbs(mappedDeal)
} catch(error) {
    if (error.cause instanceof InvalidDealError) {
        await issueAcbsFacilities(uploadDeal)
    } else {
        throw error
    }
}
```

We could even look to wrap that in it's own helper/pull it out into a service level function so 
```
    await submitDealToACBS(tfmDeal);

    if (await submitDealToACBS(tfmDeal, false, true)) {
      await issueAcbsFacilities(dealUpdate);
    }

    return tfmDeal;
```
becomes
```
await submitDealOrIssueDealFacilities(tfmDeal, dealUpdate)
```
and in a helper/service we have
```
try {
    await submitDealToAcbs(mappedDeal)
} catch(error) {
    if (error.cause instanceof InvalidDealError) {
        await issueAcbsFacilities(uploadDeal)
    } else {
        throw error
    }
}
```

tldr I'm a little worried about having a funciton saying `doSomething` and if we have `doSomething(true)` it actually does `somethingElse` and doesn't `doSomething`

Reply:
Have refactoerd. Thanks